### PR TITLE
Convert Burgers observation Action to an Event

### PR DESCRIPTION
**Pull request by wthrowe**
I intend to try to generalize this to cover all the systems in a future PR, which is why this does not touch the other observation Actions.

This uses the registration recommendations from #1328, but as that is a guideline change there is no actual code dependence between the two PRs.

Intended changes from the Action version:
* The label for the RMS error was changed from "UError" to "ErrorU" to match the volume version.
* The sign of the volume error was changed to the (I believe) more conventional value obtained by "measured - correct".

### Types of changes:

- [ ] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/Evolution/EventsAndTriggers/EventsAndTriggers.hpp
Comment (nilsvu) [inline src/Evolution/EventsAndTriggers/EventsAndTriggers.hpp:75]: #1345 is relevant here

File: src/Evolution/EventsAndTriggers/EventsAndTriggers.hpp
Comment (nilsvu) [inline src/Evolution/EventsAndTriggers/EventsAndTriggers.hpp]: Isn't the events and triggers framework independent of the concept of "slabs"?

File: src/Evolution/Executables/Burgers/EvolveBurgers.hpp
Comment (nilsvu) [inline src/Evolution/Executables/Burgers/EvolveBurgers.hpp:79]: Could you document somewhere that these type aliases are part of the public `Metavariables` interface? The `EvolutionSystemsGroup` has a list that you can extend. Then, in the events-and-triggers documentation you can refer to that and describe how the type aliases are used.

File: src/Evolution/Executables/Burgers/EvolveBurgers.hpp
Comment (nilsvu) [inline src/Evolution/Executables/Burgers/EvolveBurgers.hpp:169]: For my own education, why is this necessary?
Also, this looks like the only place where `Metavariables::events/triggers` is used.

File: src/Evolution/Systems/Burgers/Observe.hpp
Comment (nilsvu) [inline src/Evolution/Systems/Burgers/Observe.hpp:133]: remove the `using Vars` and instead inline `tmpl::list<Tags::U>{}`?

File: src/Evolution/Systems/Burgers/Observe.hpp
Comment (nilsvu) [inline src/Evolution/Systems/Burgers/Observe.hpp]: We should probably document the observed quantities in more detail in these Observe events. I suggest you make a list of volume observables and another with reduction observables in this option string.

File: tests/Unit/Evolution/Systems/Burgers/Test_Observe.cpp
Comment (nilsvu) [inline tests/Unit/Evolution/Systems/Burgers/Test_Observe.cpp]: I think disentangling this into a series of calls would make it more readable 😄

Comment (nilsvu) [inline tests/Unit/Evolution/Systems/Burgers/Test_Observe.cpp]: Why does the option string need no `:` at the end? The event's `options` list is empty, but if it were not then one would need to construct it from a dictionary following a `:`, correct?

File: tests/Unit/Evolution/Systems/Burgers/Test_Observe.cpp
Comment (nilsvu) [inline tests/Unit/Evolution/Systems/Burgers/Test_Observe.cpp:206]: I'd prefer to replace the literal `0` with some variable name that reminds the reader why it is there, e.g. `fake_singleton_id`

File: tests/Unit/Evolution/Systems/Burgers/Test_Observe.cpp
Comment (nilsvu) [inline tests/Unit/Evolution/Systems/Burgers/Test_Observe.cpp:100]: why do you need to do this here? can it move into the test or into the struct?

File: tests/Unit/Evolution/Systems/Burgers/Test_Observe.cpp
Comment (nilsvu) [inline tests/Unit/Evolution/Systems/Burgers/Test_Observe.cpp:249]: `CHECK_ITERABLE_APPROX` since the math could be different?

File: tests/Unit/Evolution/Systems/Burgers/Test_Observe.cpp
Comment (nilsvu) [inline tests/Unit/Evolution/Systems/Burgers/Test_Observe.cpp:267]: I'd prefer to compute this by hand and test for the literal value, with `approx`

File: src/Evolution/EventsAndTriggers/EventsAndTriggers.hpp
Comment (wthrowe) [reply src/Evolution/EventsAndTriggers/EventsAndTriggers.hpp]: Ugh.  Yes.  You're right.

This ties into a set of somewhat nebulous plans to reorganize a lot of this stuff that I've been thinking about, and perhaps it is time to actually deal with the problem.  All the EventsAndTrigger infrastructure is in the Evolution directory, which is wrong.  It should be moved somewhere else, but I don't know where.  (Time and Evolution should probably also be merged somehow.)  Anyway, then we can have an option tag for this somewhere in Evolution and then either a generic one for other things or specific ones for each type of problem (might just be elliptic solves).  Of course then the different tags will either have to be namespaced or have different names (which is really just fake namespacing).

I propose that for this PR I move this tag into `Time/Tags.hpp` and then file a bug where we can figure out how to reorganize the directory structure and namespaces surrounding Evolution and Time.  Does this sound like a good plan?

File: src/Evolution/Executables/Burgers/EvolveBurgers.hpp
Comment (wthrowe) [reply src/Evolution/Executables/Burgers/EvolveBurgers.hpp:79]: I do not intend these to be used outside this file and do not want them to be considered part of the interface.  They have to be public for the registration code at the end.

File: src/Evolution/Executables/Burgers/EvolveBurgers.hpp
Comment (wthrowe) [reply src/Evolution/Executables/Burgers/EvolveBurgers.hpp:169]: Something about serializing abstract base classes.  I don't know the details.

You are correct.  This is the only place (outside the metavariables struct) where those aliases are used.

File: src/Evolution/Systems/Burgers/Observe.hpp
Comment (wthrowe) [reply src/Evolution/Systems/Burgers/Observe.hpp]: Good idea.

File: src/Evolution/Systems/Burgers/Observe.hpp
Comment (wthrowe) [reply src/Evolution/Systems/Burgers/Observe.hpp:133]: Why?  That seems to just be duplicating information.

File: tests/Unit/Evolution/Systems/Burgers/Test_Observe.cpp
Comment (wthrowe) [reply tests/Unit/Evolution/Systems/Burgers/Test_Observe.cpp:100]: The language standard says it has to go here.  I do not know why that decision was made.

File: tests/Unit/Evolution/Systems/Burgers/Test_Observe.cpp
Comment (wthrowe) [reply tests/Unit/Evolution/Systems/Burgers/Test_Observe.cpp:249]: It's all integers, so I think it should be exact.  (A later version of this code has a comment about that.  I can backport it.)

File: tests/Unit/Evolution/Systems/Burgers/Test_Observe.cpp
Comment (wthrowe) [reply tests/Unit/Evolution/Systems/Burgers/Test_Observe.cpp:267]: Agreed.  This is probably too much duplication with the code being tested.

File: tests/Unit/Evolution/Systems/Burgers/Test_Observe.cpp
Comment (wthrowe) [reply tests/Unit/Evolution/Systems/Burgers/Test_Observe.cpp]: You only need the `:` if you are supplying options (although I believe a `:` with no options is fine too).  The lack of options can either be because the class takes no options or because all the options have default values and you are using those.

File: tests/Unit/Evolution/Systems/Burgers/Test_Observe.cpp
Comment (wthrowe) [reply tests/Unit/Evolution/Systems/Burgers/Test_Observe.cpp:206]: I don't see a reason to be inconsistent with what we've been doing.  If you feel strongly about this you should update the instances of this pattern that already exist.

File: tests/Unit/Evolution/Systems/Burgers/Test_Observe.cpp
Comment (wthrowe) [reply tests/Unit/Evolution/Systems/Burgers/Test_Observe.cpp:267]: Unfortunately, while this would work here, the change will not work once this is generalized to other systems in the next PR.  In that PR the variables are filled by writing to the data array of a Variables object.  Variables makes no guarantee about how that array maps to the components of the tensors, so I don't know exactly what I have initialized the tensors to.

Since I would just have to change it back to the version here in the next PR, I would prefer not to change it for this temporary specialized version.

File: src/Evolution/EventsAndTriggers/EventsAndTriggers.hpp
Comment (kidder) [reply src/Evolution/EventsAndTriggers/EventsAndTriggers.hpp]: Yes, this sounds like a reasonable plan

File: src/Evolution/Executables/Burgers/EvolveBurgers.hpp
Comment (kidder) [reply src/Evolution/Executables/Burgers/EvolveBurgers.hpp:169]: Charm++ requires derived classes to be registered in order to serialize base classes

File: tests/Unit/Evolution/Systems/Burgers/Test_Observe.cpp
Comment (kidder) [reply tests/Unit/Evolution/Systems/Burgers/Test_Observe.cpp:206]: we agree with @wthrowe

File: tests/Unit/Evolution/Systems/Burgers/Test_Observe.cpp
Comment (kidder) [reply tests/Unit/Evolution/Systems/Burgers/Test_Observe.cpp:267]: we agree with @wthrowe

**Review by kidder — COMMENTED** [review]
Group review with @nilsdeppe @carmaza @gsb76

File: src/Evolution/EventsAndTriggers/EventsAndTriggers.hpp
Comment (kidder) [inline src/Evolution/EventsAndTriggers/EventsAndTriggers.hpp:70]: move the base tag too?

**Comment by kidder** [comment]
@nilsleiffischer please look at;  I am happy with the fixups (though see the question about the base tag).

File: tests/Unit/Evolution/EventsAndTriggers/Test_EventsAndTriggers.cpp
Comment (nilsvu) [inline tests/Unit/Evolution/EventsAndTriggers/Test_EventsAndTriggers.cpp]: move to end of `#include`s as of #1370

File: src/Evolution/EventsAndTriggers/EventsAndTriggers.hpp
Comment (nilsvu) [reply src/Evolution/EventsAndTriggers/EventsAndTriggers.hpp]: Sure, sounds good 👍

File: src/Evolution/Systems/Burgers/Observe.hpp
Comment (nilsvu) [inline src/Evolution/Systems/Burgers/Observe.hpp:103]: perhaps `U - U_analytic`? Not sure about the pseudo-math here, perhaps add some text that defines `U_analytic`? (It seems important to get this OptionString right because it will set an example for those to follow.)

As a guideline we could format this as:
`  * QuantityName = Expression: Description`

File: src/Evolution/Systems/Burgers/Observe.hpp
Comment (nilsvu) [inline src/Evolution/Systems/Burgers/Observe.hpp:104]: Also add the coordinates

File: src/Evolution/Systems/Burgers/Observe.hpp
Comment (nilsvu) [inline src/Evolution/Systems/Burgers/Observe.hpp:108]: Also add `Time` and `NumberOfPoints`

File: src/Evolution/Executables/Burgers/EvolveBurgers.hpp
Comment (nilsvu) [reply src/Evolution/Executables/Burgers/EvolveBurgers.hpp:79]: could you add a comment to the `Metavariables` then? It confused me a little

File: src/Evolution/Executables/Burgers/EvolveBurgers.hpp
Comment (nilsvu) [reply src/Evolution/Executables/Burgers/EvolveBurgers.hpp:169]: thanks!

**Review by nilsvu — CHANGES_REQUESTED** [review]
please rebase and squash the current fixups 👍

File: src/Evolution/EventsAndTriggers/EventsAndTriggers.hpp
Comment (wthrowe) [reply src/Evolution/EventsAndTriggers/EventsAndTriggers.hpp:70]: This tag used by non-evolution-specific code (RunEventsAndTriggers) and we want to retrieve the E+T in the same was for all the types of problems.  We will likely want multiple EventsAndTriggers tags inheriting off this base tag, probably differing only by help string.

File: src/Evolution/Systems/Burgers/Observe.hpp
Comment (wthrowe) [reply src/Evolution/Systems/Burgers/Observe.hpp:103]: I don't understand what you are suggesting, sorry.  Could you give an example or some suggested text?

File: src/Evolution/EventsAndTriggers/EventsAndTriggers.hpp
Comment (kidder) [reply src/Evolution/EventsAndTriggers/EventsAndTriggers.hpp:70]: okay

File: src/Evolution/Systems/Burgers/Observe.hpp
Comment (nilsvu) [inline src/Evolution/Systems/Burgers/Observe.hpp:79]: If I understand correctly, we are going to be able to use this event to observe e.g. at a particular time, at every N steps, or at the final time. Therefore I think it makes more sense to split this action into _two_ events: One for reductions and one for the volume. Then we can observe reductions more frequently than the volume, for instance. What do you think @wthrowe?

Comment (wthrowe) [reply src/Evolution/Systems/Burgers/Observe.hpp:79]: Agreed.  That is actually one of the later PRs in this series.

The rough sequence I have planned is (not entirely sure how it will be split up):
1. Event for Burgers (this PR)
2. Event for arbitrary system
3. Split volume and reduction events

I may also deal with observing arbitrary fields, (as opposed to just the system variables) but it is possible that will get put off.

Comment (nilsvu) [reply src/Evolution/Systems/Burgers/Observe.hpp:79]: nice 👍 sounds like less work to split the events straight away though

Comment (wthrowe) [reply src/Evolution/Systems/Burgers/Observe.hpp:79]: Perhaps, but I'm trying to be kind to my reviewers.

Comment (nilsvu) [reply src/Evolution/Systems/Burgers/Observe.hpp:79]: I'd prefer to review the "final" (split) version, instead of the intermediate step. @kidder how about you?

Comment (kidder) [reply src/Evolution/Systems/Burgers/Observe.hpp:79]: Talking with @wthrowe , it sounds like this PR is essentially finished, and the follow-ups are almost ready (2 and 3 above) so I think it is less effort for @wthrowe to just merge this now.

Comment (nilsvu) [reply src/Evolution/Systems/Burgers/Observe.hpp:79]: It seems to me that splitting this and generalizing to an arbitrary system would revise almost all changes of the “Convert to event” commit. I don’t see the point of doing that, especially if you have these changes ready. If you don’t want to rewrite the commit order I’d be fine with just pushing them to this PR and later squash them. Then we can review the full picture, instead of code that’s going to change anyway.

Comment (nilsdeppe) [reply src/Evolution/Systems/Burgers/Observe.hpp:79]: I don't want that, I like that @wthrowe is feeding it in as smaller logical pieces. I really don't see the benefit of turning this into a monster 2000+ line PR. We're going ahead with the order @wthrowe has planned.

**Comment by nilsdeppe** [comment]
@wthrowe this'll need a rebase but overall sounds like it's ready to go :)

File: src/Evolution/Systems/Burgers/Observe.hpp
Comment (wthrowe) [reply src/Evolution/Systems/Burgers/Observe.hpp:103]: @nilsleiffischer ping

File: src/PointwiseFunctions/AnalyticSolutions/Burgers/Linear.hpp
Comment (nilsvu) [inline src/PointwiseFunctions/AnalyticSolutions/Burgers/Linear.hpp]: Why do you need these constructors?

File: src/Evolution/Systems/Burgers/Observe.hpp
Comment (nilsvu) [reply src/Evolution/Systems/Burgers/Observe.hpp:103]: If you just add the other observed quantities that I commented on below, this should be fine. I‘m just worried that other events where the quantities may be less obvious will not be sufficiently documented, but we’ll deal with that in later PRs.

File: src/PointwiseFunctions/AnalyticSolutions/Burgers/Linear.hpp
Comment (wthrowe) [reply src/PointwiseFunctions/AnalyticSolutions/Burgers/Linear.hpp]: I needed them in an early version of these commits.  I can get rid of them.  (The constructor is used to initialize the global cache in the test, but I don't need to keep a copy of the analytic solution for my own use anymore so I can change that to a move.)

**Comment by wthrowe** [comment]
I'm not sure how complex the rebase will be, so I'm going to hold off until after I squash.

File: src/Evolution/Systems/Burgers/Observe.hpp
Comment (nilsvu) [inline src/Evolution/Systems/Burgers/Observe.hpp:101]: Can you add an brief description after a `:`what „Inertial“ means here? Perhaps „The coordinates in which the Burgers equation is formulated“ (if that defines it well enough). Reason is that other systems will use multiple coordinate systems and when we output them we should make clear what they are. And „Inertial“ is not self-explanatory.

File: src/Evolution/Systems/Burgers/Observe.hpp
Comment (nilsvu) [inline src/Evolution/Systems/Burgers/Observe.hpp:107]: `=` -> `:` since this is a description

File: tests/Unit/ActionTesting.hpp
Comment (nilsvu) [inline tests/Unit/ActionTesting.hpp:967]: Why do you need this?

**Review by nilsvu — COMMENTED** [review]
You can squash

File: src/Evolution/Systems/Burgers/Observe.hpp
Comment (wthrowe) [reply src/Evolution/Systems/Burgers/Observe.hpp:101]: Explaining dual frames is not the responsibility of this class.  I would offer to link to the documentation on the subject, but there doesn't appear to be any.

File: src/Evolution/Systems/Burgers/Observe.hpp
Comment (wthrowe) [reply src/Evolution/Systems/Burgers/Observe.hpp:107]: I like the `=`.

File: tests/Unit/ActionTesting.hpp
Comment (wthrowe) [reply tests/Unit/ActionTesting.hpp:967]: Something about Charm proxies or something.  I didn't change that part of the code.  I'll check if it works with the cache const.

File: src/Evolution/Systems/Burgers/Observe.hpp
Comment (nilsdeppe) [reply src/Evolution/Systems/Burgers/Observe.hpp:101]: I agree with @wthrowe. It's likely that some of these things will be updated as dual frames are implemented. This is certainly not the place to explain dual frames/ALE/moving mesh.

File: tests/Unit/ActionTesting.hpp
Comment (nilsdeppe) [reply tests/Unit/ActionTesting.hpp:967]: I just talked with @wthrowe in the office about this. Here's what Charm++ does.

1. Entry methods should be `const` or non-`const` just like local member functions.
2. As a result, we need non-const proxies (which you can only get from a non-const cache) to invoke entry methods (sometimes).
3. This is made more complicated by Charm++ having a bug (I have not taken the time to track down where in their code) that makes groups and nodegroups be invokable always even with const proxies (or maybe it's arrays, but groups and nodegroups work correctly, I can't remember). My point is we need non-const access to the global cache.

File: src/Evolution/Systems/Burgers/Observe.hpp
Comment (kidder) [reply src/Evolution/Systems/Burgers/Observe.hpp:101]: I agree InertialCoordinates should be documented, but not here.  I see @wthrowe opened #1409 to address this, so let's move one

**Comment by wthrowe** [comment]
Rebased on develop.  There were nontrivial conflicts with #1269, so someone should take a look at the reduction type constructions.

**Review by nilsdeppe — DISMISSED** [review]
LGTM but the docs is complaining:
```
generating docs for namespac/work/spectre/src/IO/Observer/ReductionActions.hpp:70:
warning: block marked with [collect_reduction_data_tags] for \snippet should appear
twice in file Test_Observe.cpp, found it 0 times
```

**Comment by wthrowe** [comment]
Added a new commit addressing the doxygen problem.  It was a non-local effect and didn't seem to fit into any of the existing commits.

**Review by nilsdeppe — APPROVED** [review]
LGTM. Thanks @wthrowe!

**Review by kidder — APPROVED** [review]

**Comment by kidder** [comment]
@nilsleiffischer please look at

**Review by nilsvu — APPROVED** [review]

**Comment by nilsdeppe** [comment]
The GCC6 release builds timed out. I've restarted both stages in case there were outdated cache issues.

**Comment by nilsdeppe** [comment]
Tests passed, timed out during backing up of cache.

**Pull request merged**
